pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,227 @@
+		OnFailure(func(failsafe.ExecutionEvent[any]) {
+			countError.Inc()
+		}).
+		OnSuccess(func(failsafe.ExecutionEvent[any]) {
+			countSuccess.Inc()
+		}).
+		OnClose(func(event circuitbreaker.StateChangedEvent) {
+			transitionClosed.Inc()
+			level.Info(ingester.logger).Log("msg", "circuit breaker is closed", "ingester", ingesterID, "previous", event.OldState, "current", event.NewState)
+		}).
+		OnOpen(func(event circuitbreaker.StateChangedEvent) {
+			transitionOpen.Inc()
+			level.Info(ingester.logger).Log("msg", "circuit breaker is open", "ingester", ingesterID, "previous", event.OldState, "current", event.NewState)
+		}).
+		OnHalfOpen(func(event circuitbreaker.StateChangedEvent) {
+			transitionHalfOpen.Inc()
+			level.Info(ingester.logger).Log("msg", "circuit breaker is half-open", "ingester", ingesterID, "previous", event.OldState, "current", event.NewState)
+		}).
+		HandleIf(func(_ any, err error) bool { return isFailure(err) })
+
+	if cfg.testModeEnabled {
+		// In case of testing purposes, we initialize the circuit breaker with count based failure thresholding,
+		// since it is more deterministic, and therefore it is easier to predict the outcome.
+		cbBuilder = cbBuilder.WithFailureThreshold(cfg.FailureThreshold)
+	} else {
+		// In case of production code, we prefer time based failure thresholding.
+		cbBuilder = cbBuilder.WithFailureRateThreshold(cfg.FailureThreshold, cfg.FailureExecutionThreshold, cfg.ThresholdingPeriod)
+	}
+
+	cb := cbBuilder.Build()
+	return &circuitBreaker{
+		CircuitBreaker: cb,
+		ingester:       ingester,
+		executor:       failsafe.NewExecutor[any](cb),
+		startTime:      time.Now().Add(cfg.InitialDelay),
+	}
+}
+
+func isFailure(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// We only consider timeouts or ingester hitting a per-instance limit
+	// to be errors worthy of tripping the circuit breaker since these
+	// are specific to a particular ingester, not a user or request.
+
+	if errors.Is(err, context.DeadlineExceeded) {
+		return true
+	}
+
+	var ingesterErr ingesterError
+	if errors.As(err, &ingesterErr) {
+		return ingesterErr.errorCause() == mimirpb.INSTANCE_LIMIT
+	}
+
+	return false
+}
+
+func (cb *circuitBreaker) isActive() bool {
+	return cb.startTime.Before(time.Now())
+}
+
+func (cb *circuitBreaker) ingesterID() string {
+	return cb.ingester.cfg.IngesterRing.InstanceID
+}
+
+func (cb *circuitBreaker) logger() log.Logger {
+	return cb.ingester.logger
+}
+
+func (cb *circuitBreaker) metrics() *ingesterMetrics {
+	return cb.ingester.metrics
+}
+
+func (cb *circuitBreaker) config() CircuitBreakerConfig {
+	return cb.ingester.cfg.CircuitBreakerConfig
+}
+
+func (cb *circuitBreaker) get(ctx context.Context, callbackFnName string, callbackFn func() (any, error)) (any, error) {
+	res, err := cb.executor.Get(callbackFn)
+	if err != nil && errors.Is(err, circuitbreaker.ErrOpen) {
+		cb.metrics().circuitBreakerResults.WithLabelValues(cb.ingesterID(), resultOpen).Inc()
```

Comment:
Is this different than handling of failure in 

```
OnFailure(func(failsafe.ExecutionEvent[any]) {
    countError.Inc()
}).
```

case?

Reply:
It is different because `circuitbreaker.Builder`, where the `OnFailure()` is defined, will increase `countErr` when errors specified in `isFailure()` method are detected. These errors come form the ingester itself and they are ingester-specific (e.g., per-instance limits, ecc).
In this case you highlighted we want to react on an error that ingester can never produce itself, because it doesn't belong to its logic: `circuitbreaker.ErrOpen`. Hence, we need to register that error separately.

Thanks to your comment I realize that I forgot to increase `countErr` in case of reaching the push request timeout (configurable by `ingester.circuit-breaker.push-timeout`). I will add it too.